src/main/java/org/junitpioneer/jupiter/FailAtExtension.java

```diff
@@ -0,0 +1,80 @@
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import static java.lang.String.format;
+import static org.junit.jupiter.api.extension.ConditionEvaluationResult.enabled;
+import static org.junitpioneer.internal.PioneerAnnotationUtils.findClosestEnclosingAnnotation;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.Optional;
+
+import org.junit.jupiter.api.extension.ConditionEvaluationResult;
+import org.junit.jupiter.api.extension.ExecutionCondition;
+import org.junit.jupiter.api.extension.ExtensionConfigurationException;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+/**
+ * This class implements the functionality for the {@code @FailAt} annotation.
+ *
+ * @see FailAt
+ */
+class FailAtExtension implements ExecutionCondition {
+
+	private static final DateTimeFormatter ISO_8601 = DateTimeFormatter.ISO_DATE;
+
+	@Override
+	public ConditionEvaluationResult evaluateExecutionCondition(ExtensionContext context) {
+		return getFailAtDateFromAnnotation(context)
+				.map(failAtDate -> evaluateFailAtDate(context, failAtDate))
+				.orElse(enabled("No @FailAt annotation found on element"));
+	}
+
+	private Optional<LocalDate> getFailAtDateFromAnnotation(ExtensionContext context) {
+		return findClosestEnclosingAnnotation(context, FailAt.class).map(FailAt::date).map(this::parseDate);
+	}
+
+	private LocalDate parseDate(String dateString) {
+		try {
+			return LocalDate.parse(dateString, ISO_8601);
+		}
+		catch (DateTimeParseException ex) {
+			throw new ExtensionConfigurationException(
+				"The `failAtDate` string '" + dateString + "' is no valid ISO-8601 string.", ex);
+		}
+	}
+
+	private ConditionEvaluationResult evaluateFailAtDate(ExtensionContext context, LocalDate failAtDate) {
+		LocalDate today = LocalDate.now();
+		boolean isBefore = today.isBefore(failAtDate);
+
+		if (isBefore) {
+			String reportEntry = format(
+				"The `date` %s is after the current date %s, so `@FailAt` did not fails the test \"%s\". It will do so when the date is reached.",
+				failAtDate.format(ISO_8601), today.format(ISO_8601), context.getUniqueId());
+			context.publishReportEntry("FailAt", reportEntry);
+			return enabled(reportEntry);
+		} else {
+			String reportEntry = format(
+				"The current date %s is after or on the `date` %s, so `@FailAt` fails the test \"%s\". Please remove the annotation.",
+				failAtDate.format(ISO_8601), today.format(ISO_8601), context.getUniqueId());
+			context.publishReportEntry(FailAtExtension.class.getSimpleName(), reportEntry);
+
+			String message = format("The current date %s is after or on the `date` %s", today.format(ISO_8601),
+				failAtDate.format(ISO_8601));
+
+			throw new ExtensionConfigurationException(message);
```

Comment:
```suggestion
			throw new AssertionFailedError(message, failAtDate, today);
```

Or just the message - up to you.

I don't think ExtensionConfigurationException is the right exception to throw here.

Reply:
> Or just the message - up to you.
> 
> I don't think ExtensionConfigurationException is the right exception to throw here.

`AssertionFailedError` is an exception for tests by `opentest4j` (see [javadoc](https://ota4j-team.github.io/opentest4j/docs/1.0.0/api/org/opentest4j/AssertionFailedError.html) not for "production" code, so this is not usable here (couldn't be imported anyway as the dependency is has test scope).

I'm open to whatever exception we use, but I think it's some kind of configuration error, when seeing the `failAt` date is reached and therefore the extension / annotation should not be used anymore with the current configuration. I / we also use the `ExtensionConfigurationException` in other situation of this extension and the `DisabledUntil` where the date is not properly set up.

@beatngu13 What do you think?

Reply:
JUniter Jupiter (API) pulls in OpenTest4J as a [compile dependency](https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api/5.10.1), so do we? Since we are testing library, that is just fine. We already use several exception types from there:

https://github.com/search?q=repo%3Ajunit-pioneer%2Fjunit-pioneer+opentest4j+language%3AJava+path%3A%2F%5Esrc%5C%2Fmain%5C%2Fjava%5C%2Forg%5C%2Fjunitpioneer%5C%2F%2F&type=code

Reply:
Hm. Okay, will use it.